File: Server/src/main/java/dev/blackilykat/pmp/server/Playback.java
/*
 * Copyright (C) 2026 Blackilykat and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.blackilykat.pmp.server;

import dev.blackilykat.pmp.PMPConnection;
import dev.blackilykat.pmp.messages.LoginSuccessResponse;
import dev.blackilykat.pmp.messages.PlaybackControlMessage;
import dev.blackilykat.pmp.messages.PlaybackOwnershipMessage;
import dev.blackilykat.pmp.messages.PlaybackUpdateMessage;

import java.time.Instant;

/// Management of the state of playback.
public class Playback {
	/// The device which currently owns playback.
	///
	/// This is the device which is meant to be playing audio.
	///
	/// The playback owner has the final say on any playback update. Devices which do not own playback
	/// are allowed to send [PlaybackControlMessage]s but cannot directly call updates. Only the playback
	/// owner can send [PlaybackUpdateMessage]s which get broadcasted to all connected devices.
	public static Device owner = null;

	/// True if playing, false if paused. Does not get stored.
	public static boolean playing = false;

	/// If [#playing], the epoch of when playback started.
	///
	/// Else, the position in milliseconds in the track.
	public static long positionOrEpoch;

	/// Initializes saving and storing state to storage and register events related to the playback [#owner].
	public static void init() {
		positionOrEpoch = ServerStorage.MAIN.position.get();

		PMPConnection.EVENT_DISCONNECTED.register(pmpConn -> {
			if(!(pmpConn instanceof ClientConnection connection)) {
				return;
			}
			if(connection.device != owner) {
				return;
			}
			owner = null;
			if(playing) {
				playing = false;
				positionOrEpoch = Instant.now().toEpochMilli() - positionOrEpoch;
			}
			Device.broadcast(new PlaybackOwnershipMessage());
			PlaybackUpdateMessage update = new PlaybackUpdateMessage();
			update.playing = false;
			update.positionOrEpoch = positionOrEpoch;
			Device.broadcast(update);
		});

		ServerStorage.MAIN.eventMaybeSaving.register(storage -> {
			if(playing) {
				storage.markDirty();
				return;
			}
			if(positionOrEpoch != ((ServerStorage.Main) storage).position.get()) {
				storage.markDirty();
			}
		});

		ServerStorage.MAIN.eventSaving.register(s -> {
			ServerStorage.Main main = (ServerStorage.Main) s;
			if(playing) {
				main.position.set(Instant.now().toEpochMilli() - positionOrEpoch);
			} else {
				main.position.set(positionOrEpoch);
			}
		});
	}

	/// Fill fields of a [LoginSuccessResponse] with information related to playback.
	public static void fillLoginSuccessResponse(LoginSuccessResponse response) {
		response.playbackOwner = owner == null ? null : owner.id;
		response.playing = playing;
		response.track = ServerStorage.MAIN.track.get();
		response.positionOrEpoch = positionOrEpoch;
		response.positiveOptions = ServerStorage.MAIN.positiveFilterOptions.get();
		response.negativeOptions = ServerStorage.MAIN.positiveFilterOptions.get();
		response.repeat = ServerStorage.MAIN.repeat.get();
		response.shuffle = ServerStorage.MAIN.shuffle.get();
	}
}

File: AndroidClient/src/main/java/dev/blackilykat/pmp/client/android/screens/Playback.kt
/*
 * Copyright (C) 2026 Blackilykat and contributors
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <https://www.gnu.org/licenses/>.
 */

package dev.blackilykat.pmp.client.android.screens

import android.graphics.BitmapFactory
import androidx.compose.foundation.Image
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.layout.*
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.res.painterResource
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.DpSize
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.em
import dev.blackilykat.pmp.RepeatOption
import dev.blackilykat.pmp.ShuffleOption
import dev.blackilykat.pmp.client.Player
import dev.blackilykat.pmp.client.android.Mutables
import dev.blackilykat.pmp.client.android.R

@Composable
fun Playback(paddingValues: PaddingValues) {
    val playing by Mutables.playing
    val shuffle by Mutables.shuffle
    val repeat by Mutables.repeat
    val track by Mutables.currentTrack
    val albumArt by Mutables.albumArt
    val position by Mutables.position

    val artists = track?.let { track ->
        track.metadata.filter { it.key.lowercase() == "artist" }.joinToString { it.value }
    } ?: ""

    Scaffold(
        modifier = Modifier.padding(bottom = paddingValues.calculateBottomPadding()),
        bottomBar = {
            Surface(
                color = MaterialTheme.colorScheme.secondaryContainer,
            ) {
                Column {
                    Row(
                        modifier = Modifier.fillMaxWidth(),
                        horizontalArrangement = Arrangement.Center,
                        verticalAlignment = Alignment.CenterVertically,
                    ) {
                        IconButton(
                            onClick = {
                                Player.setShuffle(
                                    when (Player.getShuffle()) {
                                        ShuffleOption.ON -> ShuffleOption.OFF
                                        ShuffleOption.OFF -> ShuffleOption.ON
                                    }
                                )
                            },
                        ) {
                            when (shuffle) {
                                ShuffleOption.ON -> Icon(
                                    painter = painterResource(R.drawable.shuffle),
                                    contentDescription = "Shuffle on",
                                    modifier = Modifier.size(40.dp)
                                )

                                ShuffleOption.OFF -> Icon(
                                    painter = painterResource(R.drawable.shuffle_disabled),
                                    contentDescription = "Shuffle off",
                                    modifier = Modifier.size(40.dp)
                                )
                            }
                        }

                        Spacer(Modifier.width(10.dp))

                        IconButton(
                            onClick = { Player.previous() },
                        ) {
                            Icon(
                                painter = painterResource(R.drawable.skip_previous),
                                contentDescription = "Previous track",
                                modifier = Modifier.size(40.dp)
                            )
                        }

                        Spacer(Modifier.width(10.dp))

                        IconButton(
                            onClick = { Player.playPause() },
                            colors = IconButtonColors(
                                containerColor = MaterialTheme.colorScheme.primary,
                                contentColor = MaterialTheme.colorScheme.primaryContainer,
                                disabledContainerColor = MaterialTheme.colorScheme.primaryContainer,
                                disabledContentColor = MaterialTheme.colorScheme.primary,
                            ),
                            modifier = Modifier.size(80.dp)
                        ) {
                            if (playing) {
                                Icon(
                                    painter = painterResource(R.drawable.pause),
                                    contentDescription = "Pause",
                                    modifier = Modifier.size(40.dp)
                                )
                            } else {
                                Icon(
                                    painter = painterResource(R.drawable.play),
                                    contentDescription = "Play",
                                    modifier = Modifier.size(40.dp)
                                )
                            }
                        }

                        Spacer(Modifier.width(10.dp))

                        IconButton(
                            onClick = { Player.next() },
                        ) {
                            Icon(
                                painter = painterResource(R.drawable.skip_next),
                                contentDescription = "Next track",
                                modifier = Modifier.size(40.dp)
                            )
                        }

                        Spacer(Modifier.width(10.dp))

                        IconButton(
                            onClick = {
                                Player.setRepeat(
                                    when (Player.getRepeat()) {
                                        RepeatOption.OFF -> RepeatOption.ALL
                                        RepeatOption.ALL -> RepeatOption.TRACK
                                        RepeatOption.TRACK -> RepeatOption.OFF
                                    }
                                )
                            },
                            modifier = Modifier.size(40.dp)
                        ) {
                            when (repeat) {
                                RepeatOption.OFF -> Icon(
                                    painter = painterResource(R.drawable.repeat_off),
                                    contentDescription = "repeat off",
                                    modifier = Modifier.size(40.dp)
                                )

                                RepeatOption.ALL -> Icon(
                                    painter = painterResource(R.drawable.repeat),
                                    contentDescription = "repeat all",
                                    modifier = Modifier.size(40.dp)
                                )

                                RepeatOption.TRACK -> Icon(
                                    painter = painterResource(R.drawable.repeat_once),
                                    contentDescription = "repeat this track",
                                    modifier = Modifier.size(40.dp)
                                )
                            }
                        }
                    }

                    val seeking: MutableState<Float?> = remember { mutableStateOf(null) }

                    Row(
                        modifier = Modifier.padding(10.dp),
                        verticalAlignment = Alignment.CenterVertically
                    ) {
                        Text(formatSeconds(position / 1000))
                        Surface(
                            color = Color.Transparent,
                            modifier = Modifier.padding(5.dp).weight(1f),
                        ) {
                            val progress = (seeking.value ?: position.toFloat())
                            val max = ((track?.durationSeconds ?: 0).toFloat() * 1000)

                            LinearWavyProgressIndicator(
                                progress = { progress / max },
                                waveSpeed = WavyProgressIndicatorDefaults.LinearDeterminateWavelength / 2f,
                                amplitude = { if (playing) 0.2f else 0f },
                                modifier = Modifier.height(48.dp).padding(start = 8.dp, end = 8.dp),
                                trackColor = Color.Transparent,
                                stopSize = 0.dp,
                            )

                            Slider(
                                value = progress,
                                valueRange = 0f..max,
                                onValueChange = {
                                    seeking.value = it
                                },
                                onValueChangeFinished = {
                                    seeking.value?.let {
                                        Player.seek(it.toLong())
                                        seeking.value = null
                                    }
                                },
                                colors = SliderColors(
                                    thumbColor = SliderDefaults.colors().thumbColor,
                                    activeTrackColor = Color.Transparent,
                                    activeTickColor = Color.Transparent,
                                    inactiveTrackColor = Color.Transparent,
                                    inactiveTickColor = Color.Transparent,
                                    disabledThumbColor = Color.Transparent,
                                    disabledActiveTrackColor = Color.Transparent,
                                    disabledActiveTickColor = Color.Transparent,
                                    disabledInactiveTrackColor = Color.Transparent,
                                    disabledInactiveTickColor = Color.Transparent,
                                ),
                                thumb = {
                                    SliderDefaults.Thumb(
                                        interactionSource = remember { MutableInteractionSource() },
                                        thumbSize = DpSize(20.dp, 20.dp),
                                    )
                                }
                            )
                        }
                        Text(formatSeconds(track?.durationSeconds?.toLong() ?: 0))
                    }
                }
            }
        }
    ) { paddingValues ->
        Surface(
            color = MaterialTheme.colorScheme.secondaryContainer,
            modifier = Modifier.fillMaxSize()
        ) {
            Column(
                verticalArrangement = Arrangement.Center,
                modifier = Modifier.padding(paddingValues = paddingValues)
            ) {
                albumArt?.let {
                    Row(
                        horizontalArrangement = Arrangement.Center,
                        modifier = Modifier.fillMaxWidth()
                    ) {
                        Image(
                            bitmap = BitmapFactory.decodeByteArray(it, 0, it.size).asImageBitmap(),
                            contentDescription = "Album art",
                            modifier = Modifier.fillMaxWidth(0.7f).aspectRatio(1f).padding(bottom = 20.dp)
                        )
                    }
                }

                Text(
                    text = track?.title ?: "",
                    fontSize = 10.em,
                    textAlign = TextAlign.Center,
                    lineHeight = 1.em,
                    modifier = Modifier.fillMaxWidth().padding(
                        top = 0.dp,
                        start = 10.dp,
                        end = 10.dp,
                        bottom = 0.dp
                    )
                )

                Text(
                    text = artists,
                    fontSize = 6.em,
                    textAlign = TextAlign.Center,
                    lineHeight = 1.em,
                    modifier = Modifier.fillMaxWidth().padding(
                        top = 5.dp,
                        start = 10.dp,
                        end = 10.dp,
                        bottom = 0.dp
                    )
                )
            }
        }
    }
}

fun formatSeconds(seconds: Long): String {
    var minutes = seconds / 60;
    val seconds = seconds % 60;
    val hours = minutes / 60;
    minutes %= 60;

    return if (hours > 0) {
        String.format("%d:%02d:%02d", hours, minutes, seconds)
    } else {
        String.format("%d:%02d", minutes, seconds)
    }
}